File: fortran/src/main/java/de/irt/jfor/Fortran.java
package de.irt.jfor;

/**
 * version:     $Revision$
 * created by:  dst
 * created on:  19.10.2017 14:01
 * modified by: $Author$
 * modified on: $Date$
 */
public class Fortran {


    public static float min(double v1, double v2) {
        return (float) Math.min(v1, v2);
    }

    public static short min0(double v1, double v2) {
        return (short) Math.min(v1, v2);
    }

    public static float amin0(double v1, double v2) {
        return (short) Math.min(v1, v2);
    }

    public static int amin0(int v1, int v2) {
        return  Math.min(v1, v2);
    }

    public static short amin1(short v1, short v2) {
        return (short) Math.min(v1, v2);
    }


    public static float max(double v1, double v2) {
        return (float) Math.max(v1, v2);
    }

    public static short max0(double v1, double v2) {
        return (short) Math.max(v1, v2);
    }

    public static short amax0(float v1, float v2) {
        return (short) Math.max(v1, v2);
    }

    public static float amax1(float v1, float v2) {
        return  Math.max(v1, v2);
    }


    public static float abs(float value) {
        return  Math.abs(value);
    }

    public static short iabs(int v) {
        return (short) Math.abs(v);
    }

    public static float cabs(Complex c) {
        return (float) c.abs();
    }


    public static float sign(float v, float s) {
        return (short) dsign(v, s);
    }

    public static int sign(int v, int s) {
        v = Math.abs(v);
        return (short) (s<0 ? -v : v);
    }

    public static short isign(short v, short s) {
        return (short) sign(v, s);
    }

    public static double dsign(double v, double s) {
        v = Math.abs(v);
        return (short) (s<0 ? -v : v);
    }

    public static float sqrt(float value) {
        return (float) Math.sqrt(value);
    }

    public static double sqrt(double value) {
        return Math.sqrt(value);
    }

    public static double dsqrt(double v) {
        return Math.sqrt(v);
    }

    public static Complex csqrt(Complex value) {
        return  value.sqrt();
    }


    public static float sin(double value) {
        return (float) Math.sin(value);
    }

    public static float asin(float v) {
        return (float) Math.asin(v);
    }


    public static float cos(float v) {
        return (float) Math.cos(v);
    }

    public static float acos(float value) {
        return (float) Math.acos(value);
    }


    public static float tan(double value) {
        return (float) Math.tan(value);
    }

    public static float atan(float v) {
        return (float) Math.atan(v);
    }

    public static double datan(double v) {
        return Math.atan(v);
    }


    public static float tanh(double value) {
        return (float) Math.tanh(value);
    }

    public static float atan2(float v1, float v2) {
        return (float) Math.atan2(v1, v2);
    }

    public static double datan2(double v1, double v2) {
        return Math.atan2(v1, v2);
    }


    public static float exp(float v) {
        return (float) Math.exp(v);
    }

    public static double dexp(double v) {
        return Math.exp(v);
    }


    public static float pow(float v, float e) {
        return (float) Math.pow(v, e);
    }

    public static double pow(double v, double e) {
        return Math.pow(v, e);
    }


    public static float alog(float value) {
        return (float) Math.log(value);
    }

    public static double dlog(double v) {
        return Math.log(v);
    }

    public static float alog10(float value) {
        return (float) Math.log10(value);
    }

    public static double dlog10(double v) {
        return Math.log10(v);
    }


    public static short nint(double value) {
        return (short) Math.round(value);
    }

    public static short _int(double v) {
        return (short) v;
    }

    public static short idint(double v) {
        return (short) v;
    }

    public static float _float(short v) {
        return v;
    }

    public static float _float(double v) {
        return (short) v;
    }

    public double dble(float v) {
        return v;
    }

    public static Complex cmplx(double re, double im) {
        return Complex.of(re, im);
    }

    public static Complex cmplx(double re) {
        return cmplx(re, 0);
    }

    public static short ifix(double v) {
        return (short) (v - idint(v));
    }

    public static int len_trim(ChArr str) {
        return str.len_trim();
    }

    public Object ref(Object expr) {
        return expr;
    }
}

File: fortran/src/main/java/de/irt/kfor/Fortran.kt
package de.irt.kfor

import kotlin.math.absoluteValue
import kotlin.math.roundToInt
import kotlin.reflect.KClass


open class Fortran(val units : Units) {

    fun <U:Common> common(type : KClass<U>) : U = units.unit(type.java)

    fun <U:Fortran> function(type : KClass<U>) : U = units.unit(type.java)

    companion object {
        fun alog10(value : Double) = kotlin.math.log10(value)
        fun alog(value : Double) = kotlin.math.ln(value)
        fun sqrt(value : Double) = kotlin.math.sqrt(value)
        fun dsqrt(value : Double) = kotlin.math.sqrt(value)

        fun csqrt(v : Cplx) : Cplx = v.csqrt()
        operator fun Double.plus(v : Cplx) = v+this
        operator fun Double.minus(v : Cplx) = Cplx(this-v.re, v.im)

        fun iabs(value : Int) : Int = value.absoluteValue
        fun abs(value : Double) = value.absoluteValue
        fun abs(value : Int) : Double = value.absoluteValue.toDouble()
        fun cabs(value : Cplx) = value.cabs()

        fun idint(value : Double) : Int = value.toInt()
        fun ifix(value : Double) : Int = value.toInt()

        fun intg(value : Double) : Int = value.toInt()
        fun real(value : Int) : Double = value.toDouble()
        fun dble(value : Int) : Double = value.toDouble()
        fun dble(value : Double) : Double = value.toDouble()
        fun cplx(re : Double, im : Double) = Cplx(re, im)

        fun nint(value : Double) : Int = value.roundToInt()
        fun sign(a : Double, b: Double) : Double {
            if(b<0)
                return -b.absoluteValue
            else
                return b.absoluteValue
        }

        fun min(a : Double, b: Double) : Double = kotlin.math.min(a,b)
        fun min0(a : Int, b: Int) : Int = kotlin.math.min(a,b)
        fun amin0(a : Int, b: Int) : Double = kotlin.math.min(a,b).toDouble()
        fun amin1(a : Double, b: Double) : Double = kotlin.math.min(a,b)

        fun max(a : Double, b: Double) : Double = kotlin.math.max(a,b)
        fun max(a : Double, b: Double, c:Double) : Double = kotlin.math.max(kotlin.math.max(a,b), c)
        fun max0(a : Int, b: Int) : Int = kotlin.math.max(a,b)
        fun amax0(a : Int, b: Int) : Double = kotlin.math.max(a,b).toDouble()
        fun amax1(a : Double, b: Double) : Double = kotlin.math.max(a,b)

        fun len_trim(s : Str) : Int = s.v.trim().length
        
        fun sin(a : Double) : Double = Math.sin(a)
        fun cos(a : Double) : Double = Math.cos(a)
        fun atan2(a : Double, b: Double) : Double = Math.atan2(a,b)
        fun datan2(a : Double, b: Double) : Double = Math.atan2(a,b)
        fun atan(a : Double) : Double = Math.atan(a)
        fun tanh(a : Double) : Double = Math.tanh(a)
        fun exp(a : Double) : Double = Math.exp(a)

        fun pow(a : Double, b: Int) : Double = Math.pow(a, b.toDouble())
        fun pow(a : Double, b: Double) : Double = Math.pow(a, b)
    }
}